File: code/python/templatemethod/HouseBuilder.py
from abc import ABC, abstractmethod

class HouseBuilder(ABC):
    def build_house(self):
        self._lay_foundation()
        self.build_walls()
        self.install_roof()
        self.do_interior()

    def _lay_foundation(self):
        print("Laying the foundation with concrete and steel bars.")

    @abstractmethod
    def build_walls(self):
        pass

    @abstractmethod
    def install_roof(self):
        pass

    @abstractmethod
    def do_interior(self):
        pass

File: code/typescript/templatemethod/HouseBuilder.ts
namespace TemplateMethod {
    abstract class HouseBuilder {
        // The template method defining the sequence of steps to build a house.
        public buildHouse(): void {
            this.layFoundation();
            this.buildWalls();
            this.installRoof();
            this.doInterior();
        }

        // Common step implemented in the abstract class itself.
        private layFoundation(): void {
            console.log("Laying the foundation with concrete and steel bars.");
        }

        // Steps that need to be implemented by subclasses.
        protected abstract buildWalls(): void;

        protected abstract installRoof(): void;

        protected abstract doInterior(): void;
    }
}
